ethereum/spec/src/main/java/tech/pegasys/teku/spec/logic/versions/eip4844/forktransition/Eip4844StateUpgrade.java

```diff
@@ -0,0 +1,70 @@
+ * an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the
+ * specific language governing permissions and limitations under the License.
+ */
+
+package tech.pegasys.teku.spec.logic.versions.eip4844.forktransition;
+
+import tech.pegasys.teku.infrastructure.unsigned.UInt64;
+import tech.pegasys.teku.spec.config.SpecConfigEip4844;
+import tech.pegasys.teku.spec.datastructures.state.Fork;
+import tech.pegasys.teku.spec.datastructures.state.beaconstate.BeaconState;
+import tech.pegasys.teku.spec.datastructures.state.beaconstate.common.BeaconStateFields;
+import tech.pegasys.teku.spec.datastructures.state.beaconstate.versions.capella.BeaconStateCapella;
+import tech.pegasys.teku.spec.datastructures.state.beaconstate.versions.eip4844.BeaconStateEip4844;
+import tech.pegasys.teku.spec.logic.common.forktransition.StateUpgrade;
+import tech.pegasys.teku.spec.logic.versions.altair.helpers.BeaconStateAccessorsAltair;
+import tech.pegasys.teku.spec.schemas.SchemaDefinitionsEip4844;
+
+public class Eip4844StateUpgrade implements StateUpgrade<BeaconStateEip4844> {
+
+  private final SpecConfigEip4844 specConfig;
+  private final SchemaDefinitionsEip4844 schemaDefinitions;
+  private final BeaconStateAccessorsAltair beaconStateAccessors;
+
+  public Eip4844StateUpgrade(
+      final SpecConfigEip4844 specConfig,
+      final SchemaDefinitionsEip4844 schemaDefinitions,
+      final BeaconStateAccessorsAltair beaconStateAccessors) {
+    this.specConfig = specConfig;
+    this.schemaDefinitions = schemaDefinitions;
+    this.beaconStateAccessors = beaconStateAccessors;
+  }
+
+  @Override
+  public BeaconStateEip4844 upgrade(final BeaconState preState) {
+    final UInt64 epoch = beaconStateAccessors.getCurrentEpoch(preState);
+    BeaconStateCapella preStateCapella = BeaconStateCapella.required(preState);
+    return schemaDefinitions
+        .getBeaconStateSchema()
+        .createEmpty()
+        .updatedEip4844(
+            state -> {
+              BeaconStateFields.copyCommonFieldsFromSource(state, preState);
+
+              state.setCurrentEpochParticipation(preStateCapella.getCurrentEpochParticipation());
+              state.setPreviousEpochParticipation(preStateCapella.getPreviousEpochParticipation());
+              state.setCurrentSyncCommittee(preStateCapella.getCurrentSyncCommittee());
+              state.setNextSyncCommittee(preStateCapella.getNextSyncCommittee());
+              state.setInactivityScores(preStateCapella.getInactivityScores());
+
+              state.setFork(
+                  new Fork(
+                      preState.getFork().getCurrentVersion(),
+                      specConfig.getEip4844ForkVersion(),
+                      epoch));
+
+              state.setLatestExecutionPayloadHeader(
+                  schemaDefinitions.getExecutionPayloadHeaderSchema().getDefault());
+              state.setNextWithdrawalValidatorIndex(UInt64.ZERO);
```

Comment:
I think there should be Capella's pre `next_withdrawal_validator_index`

Reply:
you're absolutely right